File: src/lib/classes/Episode.ts
import { ExternalUrl } from "./ExternalUrl";
import { ResumePoint } from "./ResumePoint";
import { Show } from "./Show";
import { SpotifyImage } from "./SpotifyImage";

export class Episode {
  // Episode Full
  show?: Show;

  // Episode Simplified
  audio_preview_url!: string | null;
  description!: string;
  duration_ms!: number;
  explicit!: boolean;
  external_urls!: ExternalUrl;
  href!: string;
  id!: string;
  images!: SpotifyImage[];
  is_externally_hosted!: boolean;
  is_playable!: boolean;
  languages!: string[];
  name!: string;
  release_date!: string;
  release_date_precision!: string;
  resume_point!: ResumePoint;
  type!: 'episode';
  uri!: string;

  private constructor() {
  }

  public static fromFullEpisode(e: SpotifyApi.EpisodeObjectFull): Episode {
    const episode = Episode.fromSimplifiedEpisode(e);
    episode.show = Show.fromSimplifiedShow(e.show);
    return episode;
  }
  
  public static fromSimplifiedEpisode(e: SpotifyApi.EpisodeObjectSimplified): Episode {
    const episode = new Episode();
    episode.audio_preview_url = e.audio_preview_url;
    episode.description = e.description;
    episode.duration_ms = e.duration_ms;
    episode.explicit = e.explicit;
    episode.external_urls = ExternalUrl.fromExternalUrlObject(e.external_urls);
    episode.href = e.href;
    episode.id = e.id;
    episode.images = e.images.map(i => SpotifyImage.fromImageObject(i));
    episode.is_externally_hosted = e.is_externally_hosted;
    episode.is_playable = e.is_playable;
    episode.languages = e.languages;
    episode.name = e.name;
    episode.release_date = e.release_date;
    episode.release_date_precision = e.release_date_precision;
    episode.resume_point = ResumePoint.fromResumePointObject(e.resume_point);
    episode.type = e.type;
    episode.uri = e.uri;
    return episode;
  }
}

File: src/main/java/de/wittenbude/expandify/models/spotifydata/Episode.java
package de.wittenbude.expandify.models.spotifydata;


import de.wittenbude.expandify.models.spotifydata.helper.Image;
import de.wittenbude.expandify.models.spotifydata.helper.ResumePoint;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;
import se.michaelthelin.spotify.enums.ModelObjectType;
import se.michaelthelin.spotify.enums.ReleaseDatePrecision;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@Document(collection = "episode")
public class Episode {
    private String audioPreviewUrl;
    private String description;
    private Integer durationMs;
    private Boolean explicit;
    private Map<String, String> externalUrls;
    private String href;

    @Id
    private String id;

    private Image[] images;
    private Boolean isExternallyHosted;
    private Boolean isPlayable;
    private String[] languages;
    private String name;
    private String releaseDate;
    private ReleaseDatePrecision releaseDatePrecision;
    private ResumePoint resumePoint;

    @DocumentReference(lazy = true)
    private ShowSimplified show;

    private ModelObjectType type;
    private String uri;


    public Episode(se.michaelthelin.spotify.model_objects.specification.Episode episode) {
        this.audioPreviewUrl = episode.getAudioPreviewUrl();
        this.description = episode.getDescription();
        this.durationMs = episode.getDurationMs();
        this.explicit = episode.getExplicit();
        this.externalUrls = episode.getExternalUrls().getExternalUrls();
        this.href = episode.getHref();
        this.id = episode.getId();
        this.images = episode.getImages() == null ? null : Arrays.stream(episode.getImages()).map(Image::new).toArray(Image[]::new);
        this.isExternallyHosted = episode.getExternallyHosted();
        this.isPlayable = episode.getPlayable();
        this.languages = episode.getLanguages();
        this.name = episode.getName();
        this.releaseDate = episode.getReleaseDate();
        this.releaseDatePrecision = episode.getReleaseDatePrecision();
        this.resumePoint = episode.getResumePoint() == null ? null : new ResumePoint(episode.getResumePoint());
        this.show = episode.getShow() == null ? null : new ShowSimplified(episode.getShow());
        this.type = episode.getType();
        this.uri = episode.getUri();
    }
}
